diff --git a/Makefile.am b/Makefile.am
index 089d86f..d523939 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -450,20 +450,6 @@ if MCAP
 INCLUDES += -I$(builddir)/health
 endif
 
-unit_objects =
-
-unit_tests = unit/test-eir
-
-noinst_PROGRAMS += $(unit_tests)
-
-unit_test_eir_SOURCES = unit/test-eir.c src/eir.c src/glib-helper.c
-unit_test_eir_LDADD = lib/libbluetooth-private.la @GLIB_LIBS@ @CHECK_LIBS@
-unit_test_eir_CFLAGS = $(AM_CFLAGS) @CHECK_CFLAGS@
-unit_test_eir_SHORTNAME = unit
-unit_objects += $(unit_test_eir_OBJECTS)
-
-TESTS = $(unit_tests)
-
 pkgconfigdir = $(libdir)/pkgconfig
 
 pkgconfig_DATA = bluez.pc
diff --git a/configure.ac b/configure.ac
index 2f07167..50ca5dc 100644
--- a/configure.ac
+++ b/configure.ac
@@ -24,8 +24,6 @@ AC_PROG_YACC
 AM_PROG_LEX
 AM_PROG_MKDIR_P
 
-PKG_CHECK_MODULES([CHECK], [check >= 0.9.4])
-
 m4_define([_LT_AC_TAGCONFIG], [])
 m4_ifdef([AC_LIBTOOL_TAGS], [AC_LIBTOOL_TAGS([])])
 
