Index: compat-wireless-2012-02-13/net/bluetooth/hci_event.c
===================================================================
--- compat-wireless-2012-02-13.orig/net/bluetooth/hci_event.c	2012-02-14 15:01:13.853051622 -0400
+++ compat-wireless-2012-02-13/net/bluetooth/hci_event.c	2012-02-14 15:01:39.777180174 -0400
@@ -45,7 +45,7 @@
 #include <net/bluetooth/bluetooth.h>
 #include <net/bluetooth/hci_core.h>
 
-static bool enable_le;
+static bool enable_le = 1;
 
 /* Handle HCI Event packets */
 
Index: compat-wireless-2012-02-13/net/bluetooth/hci_sock.c
===================================================================
--- compat-wireless-2012-02-13.orig/net/bluetooth/hci_sock.c	2012-02-14 15:01:13.853051622 -0400
+++ compat-wireless-2012-02-13/net/bluetooth/hci_sock.c	2012-02-14 15:01:39.777180174 -0400
@@ -49,7 +49,7 @@
 #include <net/bluetooth/bluetooth.h>
 #include <net/bluetooth/hci_core.h>
 
-static bool enable_mgmt;
+static bool enable_mgmt = 1;
 
 /* ----- HCI socket interface ----- */
 
