Index: compat-wireless-2012-02-13/drivers/bluetooth/hci_h4p/core.c
===================================================================
--- compat-wireless-2012-02-13.orig/drivers/bluetooth/hci_h4p/core.c	2012-02-14 15:00:50.360935131 -0400
+++ compat-wireless-2012-02-13/drivers/bluetooth/hci_h4p/core.c	2012-02-14 15:04:13.377941836 -0400
@@ -851,7 +851,7 @@
 static int hci_h4p_hci_flush(struct hci_dev *hdev)
 {
 	struct hci_h4p_info *info;
-	info = hdev->driver_data;
+	info = hci_get_drvdata(hdev);
 
 	skb_queue_purge(&info->txq);
 
@@ -919,7 +919,7 @@
 	struct sk_buff_head fw_queue;
 	unsigned long flags;
 
-	info = hdev->driver_data;
+	info = hci_get_drvdata(hdev);
 
 	if (test_bit(HCI_RUNNING, &hdev->flags))
 		return 0;
@@ -1005,7 +1005,7 @@
 
 static int hci_h4p_hci_close(struct hci_dev *hdev)
 {
-	struct hci_h4p_info *info = hdev->driver_data;
+	struct hci_h4p_info *info = hci_get_drvdata(hdev);
 
 	if (!test_and_clear_bit(HCI_RUNNING, &hdev->flags))
 		return 0;
@@ -1027,10 +1027,6 @@
 	return 0;
 }
 
-static void hci_h4p_hci_destruct(struct hci_dev *hdev)
-{
-}
-
 static int hci_h4p_hci_send_frame(struct sk_buff *skb)
 {
 	struct hci_h4p_info *info;
@@ -1044,7 +1040,7 @@
 
 	NBT_DBG("dev %p, skb %p\n", hdev, skb);
 
-	info = hdev->driver_data;
+	info = hci_get_drvdata(hdev);
 
 	if (!test_bit(HCI_RUNNING, &hdev->flags)) {
 		dev_warn(info->dev, "Frame for non-running device\n");
@@ -1100,18 +1096,15 @@
 	info->hdev = hdev;
 
 	hdev->bus = HCI_UART;
-	hdev->driver_data = info;
+	hci_set_drvdata(hdev, info);
 
 	hdev->open = hci_h4p_hci_open;
 	hdev->close = hci_h4p_hci_close;
 	hdev->flush = hci_h4p_hci_flush;
 	hdev->send = hci_h4p_hci_send_frame;
-	hdev->destruct = hci_h4p_hci_destruct;
 	hdev->ioctl = hci_h4p_hci_ioctl;
 	set_bit(HCI_QUIRK_NO_RESET, &hdev->quirks);
 
-	hdev->owner = THIS_MODULE;
-
 	if (hci_register_dev(hdev) < 0) {
 		dev_err(info->dev, "hci_register failed %s.\n", hdev->name);
 		return -ENODEV;
